paddle/operators/detail/grpc_client.cc

```diff
@@ -0,0 +1,227 @@
+    PADDLE_ENFORCE(var);
+    // TODO(gongwb): support SelectedRows
+    PADDLE_ENFORCE(var->IsType<framework::LoDTensor>(),
+                   "Only support LoDTensor, %s has wrong type", name);
+
+    const framework::LoDTensor& tensor = var->Get<framework::LoDTensor>();
+    std::ostringstream oss;
+    framework::SerializeToStream(oss, tensor, ctx);
+    msg->set_varname(name);
+    msg->set_serialized(oss.str());
+  }
+
+  template <typename reply_t>
+  void ProcRetMsg(const reply_t& replies, int64_t idx,
+                  const framework::Scope* scope, std::string name) {}
+
+  template <typename send_t, typename recv_t>
+  void Call(grpc::CompletionQueue* cq,
+            std::vector<
+                std::unique_ptr<grpc::ClientAsyncResponseReader<recv_t>>>& rpcs,
+            ClientContext* context, SendRecvService::Stub* stub,
+            const send_t& request) {
+    rpcs.emplace_back(stub->AsyncSendVariable(context, request, cq));
+  }
+};
+
+struct GetMsg {
+  void Run(const framework::Scope* scope, const std::string name,
+           VariableMessage* msg) {
+    // FIXME(gongwb): pass device context to here.
+    msg->set_varname(name);
+  }
+
+  template <typename reply_t>
+  void ProcRetMsg(const reply_t& replies, int64_t idx,
+                  const framework::Scope* scope, std::string name) {
+    std::istringstream iss(replies[idx]->serialized());
+    framework::LoDTensor ret_tensor;
+    framework::DeserializeFromStream(iss, &ret_tensor);
+    auto* outvar = scope->FindVar(name);
+    framework::LoDTensor* out_tensor =
+        outvar->GetMutable<framework::LoDTensor>();
+    // FIXME(gongwb): do not copy.
+    auto ctx = platform::CPUDeviceContext();
+    framework::CopyFrom(ret_tensor, ctx.GetPlace(), ctx, out_tensor);
+  }
+
+  template <typename send_t, typename recv_t>
+  void Call(grpc::CompletionQueue* cq,
+            std::vector<
+                std::unique_ptr<grpc::ClientAsyncResponseReader<recv_t>>>& rpcs,
+            ClientContext* context, SendRecvService::Stub* stub,
+            const send_t& request) {
+    rpcs.emplace_back(stub->AsyncGetVariable(context, request, cq));
+  }
+};
+
+template <typename send_t, typename recv_t, typename Msg_t>
+bool AsyncGRPCClient::Call(const framework::Scope* scope,
+                           const std::vector<Var>& in,
+                           std::vector<SendStatus>& ret) {
+  grpc::CompletionQueue cq;
+  // Create a ClientContext, Status, Reply, and rpc for each backend.
+  std::vector<std::unique_ptr<SendRecvService::Stub>> stubs;
+  std::vector<std::unique_ptr<ClientContext>> contexts;
+  std::vector<std::unique_ptr<Status>> statuses;
+  std::vector<std::unique_ptr<recv_t>> replies;
+  std::vector<std::unique_ptr<grpc::ClientAsyncResponseReader<recv_t>>> rpcs;
+
+  typedef std::chrono::system_clock::time_point time_point;
+
+  for (int64_t i = 0; i < (int64_t)in.size(); i++) {
+    PADDLE_ENFORCE(channels_.find(in[i].endpoint) != channels_.end());
```

Comment:
Do not manage channels in a single client, let the caller manage this.

Reply:
Or you can write a multi-client class to do multi send.

Reply:
Done.